From 65ada0cff0b1a5c8c06846303cf2dad51980d1aa Mon Sep 17 00:00:00 2001
From: Richard Hulme <peper03@mythtv.org>
Date: Tue, 14 Apr 2015 23:44:53 +0200
Subject: [PATCH] Spawn MythFrontend from a separate thread in MythWelcome to
 avoid multiple key events being queued and restarting MythFrontend as soon as
 it exits.

Problem discussed here: http://www.gossamer-threads.com/lists/mythtv/users/585218
(cherry picked from commit 72de5f444fa094564ec6c2a136ef79c74bf2650b with a minor change to an include file to compile with QT4)
---
 mythtv/programs/mythwelcome/welcomedialog.cpp | 9 ++++++---
 mythtv/programs/mythwelcome/welcomedialog.h   | 2 +-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/mythtv/programs/mythwelcome/welcomedialog.cpp b/mythtv/programs/mythwelcome/welcomedialog.cpp
index 5b88219..3176143 100644
--- a/mythtv/programs/mythwelcome/welcomedialog.cpp
+++ b/mythtv/programs/mythwelcome/welcomedialog.cpp
@@ -8,6 +8,7 @@
 #include <QCoreApplication>
 #include <QKeyEvent>
 #include <QEvent>
+#include <qtconcurrentrun.h>
 
 // myth
 #include "exitcodes.h"
@@ -104,11 +105,14 @@ bool WelcomeDialog::Create(void)
 
 void WelcomeDialog::startFrontend(void)
 {
+    // this makes sure the button appears to click properly
+    usleep(500 * 1000);
+
     QString startFECmd = gCoreContext->GetSetting("MythWelcomeStartFECmd",
                          m_installDir + "/bin/mythfrontend");
 
     myth_system(startFECmd, kMSDisableUDPListener);
-    updateAll();
+
     m_frontendIsRunning = false;
 }
 
@@ -119,8 +123,7 @@ void WelcomeDialog::startFrontendClick(void)
 
     m_frontendIsRunning = true;
 
-    // this makes sure the button appears to click properly
-    QTimer::singleShot(500, this, SLOT(startFrontend()));
+    QtConcurrent::run(this, &WelcomeDialog::startFrontend);
 }
 
 void WelcomeDialog::checkAutoStart(void)
diff --git a/mythtv/programs/mythwelcome/welcomedialog.h b/mythtv/programs/mythwelcome/welcomedialog.h
index 880222c..a4ed292 100644
--- a/mythtv/programs/mythwelcome/welcomedialog.h
+++ b/mythtv/programs/mythwelcome/welcomedialog.h
@@ -32,7 +32,6 @@ class WelcomeDialog : public MythScreenType
 
   protected slots:
     void startFrontendClick(void);
-    void startFrontend(void);
     void updateAll(void);
     void updateStatus(void);
     void updateScreen(void);
@@ -46,6 +45,7 @@ class WelcomeDialog : public MythScreenType
     bool updateScheduledList(void);
 
   private:
+    void startFrontend(void);
     void updateStatusMessage(void);
     bool checkConnectionToServer(void);
     void checkAutoStart(void);
